cupy/cuda/memory.pyx

```diff
@@ -100,11 +101,84 @@ cdef class Memory(BaseMemory):
             self.ptr = runtime.malloc(size)
 
     def __dealloc__(self):
+        # Note: Cannot raise in the destructor! (cython/cython#1613)
         if self.ptr:
-            syncdetect._declare_synchronize()
             runtime.free(self.ptr)
 
 
+cdef inline bint is_async_alloc_supported(int device_id) except*:
+    cdef list support
+
+    if CUDA_VERSION < 11020:
+        return False
+    if not hasattr(_thread_local, 'device_support_async_alloc'):
+        # "None" for uninitialized
+        support = [None for i in range(runtime.getDeviceCount())]
+        _thread_local.device_support_async_alloc = support
+    else:
+        support = _thread_local.device_support_async_alloc
+    is_supported = support[device_id]
+    if is_supported is None:
+        is_supported = runtime.deviceGetAttribute(
+            runtime.cudaDevAttrMemoryPoolsSupported, device_id)
+        support[device_id] = is_supported
+    return is_supported
+
+
+@cython.no_gc
+cdef class MemoryAsync(BaseMemory):
+    """Memory allocation on a CUDA device.
+
+    This class provides an RAII interface of the CUDA memory allocation.
+
+    Args:
+        size (int): Size of the memory allocation in bytes.
+        stream (intptr_t): Pointer to the stream on which the memory is
+            allocated and freed.
+    """
+    cdef:
+        readonly intptr_t stream
+
+    def __init__(self, size_t size, intptr_t stream):
+        self.size = size
+        self.device_id = device.get_device_id()
+        # The stream is allowed to be destroyed before the memory is freed, so
+        # we don't need to hold a reference to the stream.
+        self.stream = stream
+        if not is_async_alloc_supported(self.device_id):
+            raise RuntimeError('Device {} does not support '
```

Comment:
Instead of doing all these checks every time we allocate memory can we cache them somehow?

Reply:
The runtime API query is cached in `is_async_alloc_supported`. But, we still need to check and raise in the constructor, as it's unclear to me if there's any way to guarantee it's working (or not) when we arrive here.

Reply:
See if 94df35a is better.

Reply:
I realized there's a couple more checks to do...So I moved all checks in one place. See if it's better.